### ISSUE TITLE
ingressgateway can't access https service with TLS termination.

### ISSUE THREAD

**Issue by savitaashture** [istio/istio#13578]
**Bug description**
I tried to set up a gateway and virtual service for an https service. 
used **SIMPLE** mode for the **TLS** termination in gateway.
And the https service return 404 Not Found.
But if i use **PASSTHROUGH** then https service return proper response.

**Expected behavior**
https request should be sent from gateway to https service with SIMPLE mode.

**Steps to reproduce the bug**
1, Install the my-nginx project from istio sample. This is a https service.
https://istio.io/docs/examples/advanced-gateways/ingress-sni-passthrough/#configure-an-ingress-gateway

2, add gateway as below
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      credentialName: "simplessl"
    hosts:
    - nginx.example.com

---

apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: nginx
spec:
  hosts:
  - nginx.example.com
  gateways:
  - mygateway
  tls:
  - match:
    - port: 443
      sni_hosts:
      - nginx.example.com
    route:
    - destination:
        host: my-nginx
        port:
          number: 443
```
3. run curl command
```
curl https://nginx.example.com -k -v
```
run result
```
* Rebuilt URL to: https://nginx.example.com/
*   Trying 104.42.42.106...
* Connected to nginx.example.com (104.42.42.106) port 443 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 594 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: nginx.example.com (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: C=US,ST=Denial,L=Springfield,O=Dis,CN=nginx.example.com
*        start date: Wed, 24 Apr 2019 10:08:21 GMT
*        expire date: Sun, 03 May 2020 10:08:21 GMT
*        issuer: C=US,ST=Denial,O=Dis,CN=nginx.example.com
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET / HTTP/1.1
> Host: nginx.example.com
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< location: https://nginx.example.com/
< date: Wed, 24 Apr 2019 10:22:18 GMT
< server: istio-envoy
< content-length: 0
```

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
istioctl version
```
version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.1"}
```

kubectl version

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.5", GitCommit:"51dd616cdd25d6ee22c83a858773b607328a18ec", GitTreeState:"clean", BuildDate:"2019-01-16T18:14:49Z", GoVersion:"go1.10.7", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**

```
for i in install/kubernetes/helm/istio-init/files/crd*yaml; do kubectl apply -f $i; done
```
```
kubectl apply -f install/kubernetes/istio-demo.yaml
```
Configure a TLS ingress gateway using SDS
```
$ helm template install/kubernetes/helm/istio/ --name istio \
--namespace istio-system -x charts/gateways/templates/deployment.yaml \
--set gateways.istio-egressgateway.enabled=false \
--set gateways.istio-ingressgateway.sds.enabled=true > \
$HOME/istio-ingressgateway.yaml
$ kubectl apply -f $HOME/istio-ingressgateway.yaml
```

**Environment where bug was observed (cloud vendor, OS, etc)**
```
Linux aks-agentpool-14689735-0 4.15.0-1037-azure #39~16.04.1-Ubuntu SMP Tue Jan 15 17:20:47 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by yangminzhu** [comment]
Are you only enabling the SDS on ingress but not in the mesh? /cc @JimmyCYJ

**Comment by savitaashture** [comment]
@yangminzhu 
i followed these steps
https://istio.io/docs/tasks/traffic-management/secure-ingress/sds/

**Comment by JimmyCYJ** [comment]
From the curl response, it looks like the SSL handshake is okay, but ingress gateway could not find resource for the request. Probably the VirtualService tls section is not configured correctly. Could you try with the file mount based approach? https://istio.io/docs/tasks/traffic-management/secure-ingress/mount/
If the issue still exists, then we can check VirtualService configuration or ask networking team to take a look.

**Comment by dbarentine** [comment]
Same issue here only my backend destination is just HTTP and I'm running Istio 1.1.5 and Kubernetes 1.12.5. I'm terminating TLS at the ingress gateway which is using SDS. I'm only using SDS for ingress.

It's also not a consistent 404 across the board. I have one service where I get a 404 in Chrome but not in Firefox. One that seems to work fine all the time currently. If I add or remove a gateway then the issues seem to shuffle. But it's consistent once the gateway definitions are in place. Curl never seems to have an issue.

HTTP requests all work with no issues. It's only ingressgateway terminated SSL connections. 

I can give the file mount approach a try tomorrow and see if it makes any difference.

**Comment by dbarentine** [comment]
I just tried the file mount approach and it exhibits the same behavior.

**Comment by savitaashture** [comment]
@JimmyCYJ 

I have used file mount based approach
https://istio.io/docs/tasks/traffic-management/secure-ingress/mount/

**Gateway**
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - httpbin.example.com
```

**VirtualService**

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: nginx
spec:
  hosts:
  - httpbin.example.com
  gateways:
  - mygateway
  tls:
  - match:
    - port: 443
      sni_hosts:
      - httpbin.example.com
    route:
    - destination:
        host: my-nginx
        port:
          number: 443
```

Still getting 404 error

```
curl -v --resolve httpbin.example.com:$SECURE_INGRESS_PORT:$INGRESS_HOST https://httpbin.example.com:$SECURE_INGRESS_PORT  --cacert httpbin.example.com/2_intermediate/certs/ca-chain.cert.pem
* Added httpbin.example.com:443:10.100.152.228 to DNS cache
* Rebuilt URL to: https://httpbin.example.com:443/
* Hostname httpbin.example.com was found in DNS cache
*   Trying 10.100.152.228...
* Connected to httpbin.example.com (10.100.152.228) port 443 (#0)
* found 2 certificates in httpbin.example.com/2_intermediate/certs/ca-chain.cert.pem
* found 692 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification OK
*        server certificate status verification SKIPPED
*        common name: httpbin.example.com (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: C=US,ST=Denial,L=Springfield,O=Dis,CN=httpbin.example.com
*        start date: Wed, 08 May 2019 04:02:35 GMT
*        expire date: Sun, 17 May 2020 04:02:35 GMT
*        issuer: C=US,ST=Denial,O=Dis,CN=httpbin.example.com
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET / HTTP/1.1
> Host: httpbin.example.com
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< location: https://httpbin.example.com/
< date: Wed, 08 May 2019 04:14:38 GMT
< server: envoy
< content-length: 0
```

**Comment by JimmyCYJ** [comment]
Thanks @savitaashture, it shows that SSL handshake has finished, but Envoy cannot find backend. The 404 is returned from here https://github.com/envoyproxy/envoy/blob/a79164f50f55cfd02f0516eeb693c8fee18e0f90/source/common/router/config_utility.cc#L114

I suspect the following part is not set up correct. @PiotrSikora could you help to take a look?
 tls:
  - match:
    - port: 443
      sni_hosts:

**Comment by savitaashture** [comment]
Same **tls** is working if mode is **PASSTHROUGH**

**Comment by JimmyCYJ** [comment]
Tentatively assign to @PiotrSikora as I think the ingress gateway SDS works fine here. SSL connection has established successfully, which does not cause 404. It seems that the 404 is returned by routing rules. https://github.com/istio/istio/issues/13578#issuecomment-490351516

I could be wrong, please assign back to me if it is ingress SDS issue.

**Comment by PiotrSikora** [comment]
I believe the issue is that you're using example config for  **PASSTHROUGH** mode, and expecting it to work in **SIMPLE** mode without any changes.

Notably, you're sending TLS-terminated traffic to port 443:
```
- destination:
    host: my-nginx
    port:
      number: 443
```
where NGINX expects TLS traffic, instead of port 80.

If you look at the response, it actually tries to redirect you to HTTPS using `Location` header.
```
< HTTP/1.1 404 Not Found
< location: https://httpbin.example.com/
< date: Wed, 08 May 2019 04:14:38 GMT
< server: envoy
< content-length: 0
```
(I'm not sure why the response has `404 Not Found` status instead of `301 Permanent Redirect`, though, there might be a bug somewhere...)

**Comment by hanoisteve** [comment]
So is this working now?

**Comment by vadimeisenbergibm** [comment]
@savitaashture cc @PiotrSikora Note that when you use TLS termination in the ingress Gateway, you then have to treat the traffic as HTTP in your VirtualService. See the example in https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-mount/#configure-a-tls-ingress-gateway-with-a-file-mount-based-approach. Since the destination service expects HTTPS, you have to define a DestinationRule so that TLS origination will be performed before calling the destination service.

To summarize, the traffic schematically flows this way:
client sends HTTPS ->  TLS termination by the ingress gateway -> HTTP -> HTTP routing by the VirtualService -> TLS origination by the DestinationRule sends HTTPS -> destination service.

**Comment by eyeezzi** [comment]
@vadimeisenbergibm, can you give example YAML files for this TLS origination you talk about. I couldn't find any in the documentation you linked.

**Comment by vadimeisenbergibm** [comment]
@eyeezzi The example configuration is described in https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-mount/#configure-a-tls-ingress-gateway-with-a-file-mount-based-approach. Ingress Gateway performs TLS termination. The (mutual) TLS origination is performed by default when it is enabled in the mesh. To enable TLS origination inside the mesh - https://istio.io/docs/tasks/security/authn-policy/.

**Comment by eyeezzi** [comment]
@vadimeisenbergibm Thanks for the resources, but in my case I want the ingress-gateway to behave as follows: 
1. Terminate Simple TLS connection from the browser.
2. Then originate Simple TLS to an upstream service **outside the mesh**. 

## For context, this is my use case:

1. My upstream service: `kiali.istio-system.svc.cluster.local` is a server (without a sidecar) that accepts only Simple TLS connections on TCP port 20001. It has a *cert.pem* and *key.pem* file.
2. The ingress-gateway is setup to terminate Simple TLS connection with browsers. It is mounted a `serverCertificate: tls.crt` and `privateKey:tls.key` file.
3. There's a VirtualService on the ingress-gateway configured to route all traffic where `host=kiali.domain.com` to the upstream `kiali.istio-system.svc.cluster.local`.
4. There's a DestinationRule to force Simple TLS when connecting to the upstream `kiali.istio-system.svc.cluster.local`.

I haven't found any reference for this particular setup, so I'm just patching up YAMLs in hopes that it works...it doesn't and that's why I'm here.

## Here are my config files:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: internet-to-igw
  namespace: default
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - kiali.domain.com
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt # exists, confirmed
      privateKey: /etc/istio/ingressgateway-certs/tls.key # exists, confirmed
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: igw-to-kiali-https
spec:
  hosts:
  - kiali.domain.com
  gateways:
  - internet-to-igw
  http:
  - match:
    route:
    - destination:
        host: kiali.istio-system.svc.cluster.local
        port:
          number: 20001
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: originate-tls-for-kiali
spec:
  host: kiali.istio-system.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 20001
      tls:
        mode: SIMPLE
```
## My debugging efforts

- Connecting from browser to internet gateway (ignoring certificate verification). TLS termination works fine, but ingress-gateway wrongly attempts to connect to upstream with HTTP..and fails.
```
curl -k -v https://kiali.domain.com

* Rebuilt URL to: https://kiali.domain.com/
*   Trying {IP}...
* TCP_NODELAY set
* Connected to kiali.domain.com ({IP}) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=Denial; L=Springfield; O=Dis; CN={...}
*  start date: Aug  8 22:47:07 2019 GMT
*  expire date: Aug 17 22:47:07 2020 GMT
*  issuer: C=US; ST=Denial; O=Dis; CN={...}
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fb31a005200)
> GET / HTTP/2
> Host: kiali.domain.com
> User-Agent: curl/7.54.0
> Accept: */*
> 
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 400 
< x-envoy-upstream-service-time: 1
< date: Mon, 12 Aug 2019 02:41:04 GMT
< server: istio-envoy
< 
Client sent an HTTP request to an HTTPS server.
* Connection #0 to host kiali.domain.com left intact
```
* Accessing the upstream service directly from the ingress-gateway container. It appears the ingress-gateway is failing to verify the upstream's certificate, because when curl works if I ignore certificate verification (with the `-k` flag)
```
kubectl -n istio-system exec -it $(kubectl -n istio-system get pod -l app=istio-ingressgateway -o jsonpath={.items..metadata.name}) -c istio-proxy -- curl -v https://kiali.istio-system.svc.cluster.local:20001

* Rebuilt URL to: https://kiali.istio-system.svc.cluster.local:20001/
*   Trying 10.12.11.96...
* Connected to kiali.istio-system.svc.cluster.local (10.12.11.96) port 20001 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 592 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
* server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none
* Closing connection 0
curl: (60) server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none
More details here: http://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
command terminated with exit code 60
```

Sorry if this is long, I just want to put the details here for completeness. In summary, my question is:
1. How do I configure the ingress-gateway to terminate TLS, and originate TLS for the upstream out-of-mesh service?
2. How do I configure the ingress-gateway to trust the certificate of the upstream out-of-mesh service?

If you read all that, much thanks.

**Comment by vadimeisenbergibm** [comment]
To handle 404, I think you should add `uri` under `match`:

```
- match:
    - uri:
        prefix: /
```

To perform TLS origination to all your HTTPS service, you will have to mount the certificate of the service as another certificate to your ingress gateway, similar to the following task:

https://istio.io/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#redeploy-the-egress-gateway-with-the-client-certificates

**Comment by gustavovalverde** [comment]
I'm having the same problem; I'm getting 404 when using HTTPS. Port 80 working great.

#### Config files

```yaml
apiVersion: v1
kind: Service
metadata:
  name: odoo-service
  labels:
    app: odoo
    istio: ingressgateway
  namespace: default
spec:
  selector:
    app: odoo
  ports:
  - name: https-odoo
    port: 443
    targetPort: 8069
  - name: http-odoo
    port: 8069
    targetPort: 8069
  - name: http-longpolling
    port: 8072
    targetPort: 8072
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: istio-odoo
spec:
  gateways:
  - istio-odoo.default
  hosts:
  - istio.iterativo.do
  http:
  - match:
    - uri:
      prefix: /
    route:
    - destination:
        host: odoo-service.default.svc.cluster.local
        port:
          number: 8069
  - match:
    - uri:
      prefix: /longpolling
    route:
    - destination:
        host: odoo-service.default.svc.cluster.local # Destination service, must match service name
        port:
          number: 8072
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-odoo # Individual name (we'll use this later in VirtualServer)
  namespace: default # Namespace
spec:
  selector:
    istio: ingressgateway # Specify the ingressgateway created for us by GKE
  servers:
  - hosts:
    - '*'
    port:
      name: http
      number: 80
      protocol: HTTP2
  - hosts:
    - '*'
    port:
      name: https-default
      number: 443
      protocol: HTTPS
    tls:
      mode: SIMPLE
      privateKey: /etc/istio/ingressgateway-certs/tls.key
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
```

#### Curl result
```bash
* Rebuilt URL to: https://istio.iterativo.do/
*   Trying 34.74.1.138...
* TCP_NODELAY set
* Connected to istio.iterativo.do (34.74.1.138) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=iterativo.do
*  start date: Aug 19 03:06:20 2019 GMT
*  expire date: Nov 17 03:06:20 2019 GMT
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55ea197b7e80)
> GET / HTTP/1.1
> Host: istio.iterativo.do
> User-Agent: curl/7.52.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 404
< location: https://istio.iterativo.do/
< date: Mon, 19 Aug 2019 05:43:39 GMT
< server: istio-envoy
<
* Curl_http_done: called premature == 0
* Connection #0 to host istio.iterativo.do left intact
```

Any updates on what I might be doing wrong?

**Comment by vadimeisenbergibm** [comment]
@gustavovalverde Can you rename?
```
- name: 
https-odoo
    port: 8069
    targetPort: 8069
```

I guess the port should be `http-something-odoo`, it is not HTTPS. The HTTPS in the port name that you send HTTP traffic probably confuses Istio.

Also, what is the result of sending a request to `https://istio.iterativo.do/longpolling`?

**Comment by gustavovalverde** [comment]
@vadimeisenbergibm I actually tried two things. I renamed the **https** port to another thing and I also completely removed the `https` port as it is sending to the same port as the `http `one. Same result.

With `https://istio.iterativo.do/longpolling` I have the same result; even though is working with `http`

**Comment by vadimeisenbergibm** [comment]
Two more things to try:
1. Indent `prefix` to the right, under `uri`
2. Set the hosts in the gateway to `istio.iterativo.do`

**Comment by vadimeisenbergibm** [comment]
Another thing: try to remove `istio: ingressgateway` from your Service definition.

**Comment by gustavovalverde** [comment]
That's curious. I tested changing (1) The `prefix` indent and (2) Removing the `istio: ingressgateway`. One of those things made it work; too bad I did it at the same time.

Reverting back both changes is not causing the 404 anymore (Is there some caching mechanism with the rules in the Istio proxy?)

**Comment by vadimeisenbergibm** [comment]
@gustavovalverde To be on the safe side, delete the virtual service and verify that the routing stopped working. Then define the virtual service again.

**Comment by gustavovalverde** [comment]
@vadimeisenbergibm well it seems like something was "hanging" there. After deleting and re-testing everything, the culprit was:
- Indent `prefix `to the right, under `uri `

I'm a bit lost why this caused the port 80 to work but not 443.

**Comment by vadimeisenbergibm** [comment]
@gustavovalverde OK, I am glad it works.

**Comment by gustavovalverde** [comment]
@vadimeisenbergibm Thank you very much for your time and patience.

**Comment by vadimeisenbergibm** [comment]
@B-Hancock the wildcard (`*`) can be only at the beginning of the host name, before a dot. Please try `*.io`.

**Comment by vadimeisenbergibm** [comment]
@B-Hancock Oh, I see. 

I would check if there are some other gateways or virtual services are defined for the same host, maybe there is some conflict. See  https://istio.io/docs/ops/traffic-management/troubleshooting/#route-rules-have-no-effect-on-ingress-gateway-requests.

**Comment by vadimeisenbergibm** [comment]
@B-Hancock OK, to debug the issue, can you dump the working and non-working configuration, and compare them? See https://istio.io/docs/ops/component-debugging/#get-proxy-configuration

**Comment by ishanrakitha** [comment]
Hi,

I'm also facing the same issue. Port 80 is working fine. 443 get a 404. Domain is on cloud-flare. SDS Certificates are in place.  Could you please advise on the below. Appreciate your support in advance. 

`info	sdsServiceLog	CONNECTION ID: router~10.244.0.107~istio-ingressgateway-b6cc895cd-7c42n.istio-system~istio-system.svc.cluster.local-9, RESOURCE NAME: test-certificate, EVENT: pushed key/cert pair to proxy`

`apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: nginx-gateway
  namespace: istio-system
  labels:
    app: ingressgateway
spec:
  selector:
    istio: ingressgateway
  servers:
    - port:
        number: 443
        protocol: HTTPS
        name: https
      tls:
        mode: SIMPLE
        serverCertificate: "sds"
        privateKey: "sds"
        credentialName: "test-certificate"
      hosts:
        - "$DOMAIN_NAME"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: nginx-vs
  namespace: istio-system
spec:
  hosts:
    - "$DOMAIN_NAME"
  gateways:
    - nginx-gateway
  http:
    - match:
        - port: 443
      route:
        - destination:
            host: nginx-service
            port:
              number: 80
EOF`

Thanks & Regards,
Ishan.

**Comment by vadimeisenbergibm** [comment]
@ishanrakitha According to https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-sds/#configure-a-tls-ingress-gateway-for-multiple-hosts, you should only use `credentialName`, not the serverCertificate and privateKey, as appears in your configuration.

Does https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-sds/#configure-a-tls-ingress-gateway-for-multiple-hosts work for you?

**Comment by Baitanik** [comment]
I am facing the same issue... 
```
    name: istio-gateway
    namespace: istio-system
  spec:
    selector:
      istio: ingressgateway
    servers:
    - hosts:
      - '*'
      port:
        name: https
        number: 8443
        protocol: HTTPS
      tls:
        mode: SIMPLE
        privateKey: /etc/istio/ingressgateway-certs/tls.key
        serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
    - hosts:
      - '*'
      port:
        name: http
        number: 80
        protocol: HTTP
```
virtual service:
```
apiVersion: networking.istio.io/v1beta1
  kind: VirtualService
  metadata:
    name: metalk8s-ui-proxies-https
    namespace: metalk8s-ui
  spec:
    gateways:
    - istio-system/istio-gateway
    hosts:
    - '*'
    http:
    - match:
      - uri:
          prefix: /api/kubernetes/
      route:
      - destination:
          host: kubernetes-api-ds
          port:
            number: 443
        weight: 100


```

destination rule:
```
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: kubernetes-api-ds
  namespace: metalk8s-ui
spec:
  host: kubernetes-api
  trafficPolicy:
    portLevelSettings:
    - loadBalancer:
        simple: ROUND_ROBIN
      port:
        number: 443
    tls:
      caCertificates: /etc/istio/ingressgateway-certs/tls.crt
      mode: SIMPLE
      privateKey: /etc/istio/ingressgateway-certs/tls.key

```

when i am trying to call ... it is giving 503 error... 
```
Request URL: https://10.105.58.133:8443/api/kubernetes/api/v1/namespaces?includeUninitialized=&pretty=&continue=&fieldSelector=metadata.name%3Dkube-system
Request Method: GET
Status Code: 503 
Remote Address: 10.105.58.133:8443
Referrer Policy: strict-origin-when-cross-origin
```
could anyone please help

**Comment by yonwon01** [comment]
> @vadimeisenbergibm The code I have above is actually running in a test environment so that I can rule out issues like that. This is the only gateway/vs.

sorry, but Did you solve the problem??

**Comment by VadimEisenberg** [comment]
@yonwon01 I do now work on Istio starting from 2020, sorry. The issue was closed at some point.
